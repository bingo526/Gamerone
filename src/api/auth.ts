import request, { stringifyBody } from 'utils/request';
import {
  LoginRequest,
  LoginResponse,
  SignupRequest,
  CheckEmailRequest,
  CheckUsernameRequest,
  VerifyTokenRequest,
  StatusResponse,
  User,
} from 'interfaces';

const requestOptions = (param: any): RequestInit => ({
  method: 'POST',
  body: stringifyBody(param),
});

/**
 * Sign up as a new user
 */
export const signUp = (param: SignupRequest) => {
  return request<LoginResponse>('/signup', requestOptions(param));
};

/**
 * Check for unique email address
 */
export const checkEmail = (param: string) => {
  return request<StatusResponse>(
    '/check-email',
    requestOptions({ email: param } as CheckEmailRequest),
  );
};

/**
 * Check for unique username
 */
export const checkUsername = (param: string) => {
  return request<StatusResponse>(
    '/check-username',
    requestOptions({ username: param } as CheckUsernameRequest),
  );
};

/**
 * Log in with an existing user
 */
export const login = (param: LoginRequest) => {
  return request<LoginResponse>('/login', requestOptions(param));
};

/**
 * Verify email with token
 */
export const verifyEmail = (param: VerifyTokenRequest) => {
  return request<StatusResponse>('/verify-token', requestOptions(param));
};

/**
 * Get user information for authenticated user
 */
export const getMe = () => {
  return request<User>('/me');
};
