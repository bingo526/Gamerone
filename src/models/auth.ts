import { LoginRequest, SignupRequest, AcceptedEnum } from 'interfaces';

export class LoginActionPayload {
  prevPath = '/';
  request: LoginRequest = {
    email: '',
    password: '',
  };

  fromRequest = (request: LoginRequest, from?: { pathname: string }): this => {
    this.request = request;
    this.prevPath = from ? from.pathname : '/';
    return this;
  };
}

export class SignupActionPayload implements SignupRequest {
  email = '';
  password = '';
  username = '';
  firstName = '';
  lastName = '';
  accepted = AcceptedEnum.NUMBER_0;

  fromRequest = (request: SignupRequest, accepted: boolean): this => {
    Object.assign(this, request);
    if (accepted) {
      this.accepted = AcceptedEnum.NUMBER_1;
    } else {
      this.accepted = AcceptedEnum.NUMBER_0;
    }
    return this;
  };
}
