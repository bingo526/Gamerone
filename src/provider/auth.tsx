import React, { createContext } from 'react';

interface Props {
  isAuthenticated: boolean;
  children?: React.ReactNode;
}
export const AuthContext = createContext({
  isAuthenticated: false,
});

export function AuthProvider({ isAuthenticated, children }: Props) {
  return (
    <AuthContext.Provider value={{ isAuthenticated }}>
      {children}
    </AuthContext.Provider>
  );
}
