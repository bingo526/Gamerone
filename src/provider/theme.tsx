import React from 'react';
import { G1_THEMES } from 'utils/constants';

export type G1ThemeName = keyof typeof G1_THEMES;
export type G1ThemeValue = typeof G1_THEMES[G1ThemeName];

type Action =
  | { type: 'SET_THEME'; payload: G1ThemeName }
  | { type: 'SET_SYSTEM_DEFAULT'; payload: boolean }
  | { type: 'SET_DARK_MODE'; payload: boolean };

type Dispatch = (action: Action) => void;
type State = {
  theme: G1ThemeName;
  isSystemDefault: boolean;
  isDarkMode: boolean;
};

type ThemeProviderProps = { children: React.ReactNode };

const ThemeStateContext = React.createContext<State | undefined>(undefined);
const ThemeDispatchContext = React.createContext<Dispatch | undefined>(
  undefined,
);

function themeReducer(state: State, action: Action) {
  switch (action.type) {
    case 'SET_THEME': {
      localStorage.setItem('theme', action.payload);
      return { ...state, theme: action.payload };
    }

    case 'SET_SYSTEM_DEFAULT': {
      localStorage.setItem('systemDefault', action.payload ? 'on' : 'off');
      return { ...state, isSystemDefault: action.payload };
    }

    case 'SET_DARK_MODE': {
      localStorage.setItem('darkMode', action.payload ? 'on' : 'off');
      return { ...state, isDarkMode: action.payload };
    }

    default: {
      throw new Error(`Unhandled action type: ${action}`);
    }
  }
}

function ThemeProvider({ children }: ThemeProviderProps) {
  const systemDefault = localStorage.getItem('systemDefault');
  const darkMode = localStorage.getItem('darkMode');
  const savedTheme = localStorage.getItem('theme') || 'Default';

  if (!Object.keys(G1_THEMES).includes(savedTheme)) {
    throw new Error('Unidentifed theme name');
  }

  const [state, dispatch] = React.useReducer(themeReducer, {
    theme: savedTheme as G1ThemeName,
    isSystemDefault: systemDefault
      ? systemDefault === 'on'
        ? true
        : false
      : true,
    isDarkMode: darkMode ? (darkMode === 'on' ? true : false) : false,
  });

  return (
    <ThemeStateContext.Provider value={state}>
      <ThemeDispatchContext.Provider value={dispatch}>
        {children}
      </ThemeDispatchContext.Provider>
    </ThemeStateContext.Provider>
  );
}

function useThemeState() {
  const context = React.useContext(ThemeStateContext);
  if (context === undefined) {
    throw new Error('useThemeState must be used within a ThemeProvider');
  }

  return context;
}

function useThemedispatch() {
  const context = React.useContext(ThemeDispatchContext);

  if (context === undefined) {
    throw new Error('useThemeDispatch must be used within a ThemeProvider');
  }

  return context;
}

export { ThemeProvider, useThemeState, useThemedispatch };
