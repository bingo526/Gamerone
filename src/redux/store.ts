import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import { createBrowserHistory as createHistory } from 'history';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import thunk from 'redux-thunk';
import createSagaMiddleware from 'redux-saga';
import reducers from './reducers';
import rootSaga from './sagas';

const history = createHistory();
const sagaMiddleware = createSagaMiddleware();
const routeMiddleware = routerMiddleware(history);
const middlewares = [thunk, sagaMiddleware, routeMiddleware];

const composeEnhancers =
  (window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] as typeof compose) || compose;

const store = createStore(
  combineReducers({
    ...reducers,
    router: connectRouter(history),
  }),
  composeEnhancers(applyMiddleware(...middlewares)),
);
sagaMiddleware.run(rootSaga);
export { store, history };
