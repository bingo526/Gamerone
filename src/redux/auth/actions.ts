import {
  INIT_AUTH,
  CHECK_AUTHORIZATION,
  LOGIN_REQUEST,
  SIGNUP_SUCCESS,
  LOGOUT,
  SIGNUP_REQUEST,
  UPDATE_CURRENT_USER,
  GET_ME,
  AuthActionTypes,
} from './types';

import { User, SignupResponse } from 'interfaces';
import { LoginActionPayload, SignupActionPayload } from 'models/auth';

export default {
  initAuth: (): AuthActionTypes => ({ type: INIT_AUTH }),
  getMe: (): AuthActionTypes => ({ type: GET_ME }),

  checkAuthorization: (): AuthActionTypes => ({
    type: CHECK_AUTHORIZATION,
  }),
  login: (payload: LoginActionPayload): AuthActionTypes => ({
    type: LOGIN_REQUEST,
    payload,
  }),
  logout: (): AuthActionTypes => ({
    type: LOGOUT,
  }),

  signUp: (payload: SignupActionPayload): AuthActionTypes => ({
    type: SIGNUP_REQUEST,
    payload,
  }),
  signUpSuccess: (payload: SignupResponse): AuthActionTypes => ({
    type: SIGNUP_SUCCESS,
    payload,
  }),

  updateCurrentUser: (payload: User): AuthActionTypes => ({
    type: UPDATE_CURRENT_USER,
    payload,
  }),
};
