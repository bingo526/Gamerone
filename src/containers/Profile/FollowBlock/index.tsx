/* eslint-disable react-hooks/exhaustive-deps */
import React, { memo, useContext, useMemo, useCallback } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';

import Button, { ButtonSchemeEnum } from 'components/common/Button';
import { selectCurrentProfileUser } from 'redux/profile/selectors';
import ProfileActions from 'redux/profile/actions';
import SettingsActions from 'redux/settings/actions';
import { AuthContext } from 'provider/auth';
import { RootState } from 'redux/types';
import { UserModel } from 'models/user';
import Dropdown, { DropdownOptionType } from 'components/common/Dropdown';
import { UserSquadStatusEnum } from 'interfaces';

export interface FollowBlockProps {
  isSelf: boolean;
  className?: string;
}

function FollowBlock({
  isSelf,
  user,
  dispatchFollow,
  dispatchUnfollow,
  dispatchBlock,
  dispatchAddSquad,
  dispatchRemoveSquad,
}: FollowBlockProps & ConnectedProps) {
  const { isAuthenticated } = useContext(AuthContext);
  let { isFollowing } = (user as UserModel).followStatus;
  const { isBlocking = false } = (user as UserModel).followStatus;

  const handleFollowClick = useCallback(
    (e: React.MouseEvent<HTMLButtonElement>) => {
      e.preventDefault();
      if (!isFollowing) {
        dispatchFollow();
        isFollowing = true;
      }
    },
    [isFollowing],
  );

  const handleUnfollowClick = useCallback(() => {
    if (isFollowing) dispatchUnfollow();
  }, [isFollowing]);

  const dropdownOptions = useMemo(() => {
    const options: DropdownOptionType[] = [];

    if (isFollowing) {
      options.push({
        label: 'Unfollow',
        onClick: (): void => handleUnfollowClick(),
      });
    }
    if (user?.isSquad === UserSquadStatusEnum.NO_SQUAD) {
      options.push({
        label: 'Invite to Squad',
        onClick: () => dispatchAddSquad(),
      });
    } else if (user?.isSquad === UserSquadStatusEnum.IN_SQUAD) {
      options.push({
        label: 'Remove from Squad',
        onClick: () => dispatchRemoveSquad(user.id),
      });
    } else {
      options.push({ label: 'Waiting for accept', disabled: true });
    }

    options.push({ label: 'Block', onClick: () => dispatchBlock() });
    options.push({ label: 'Unblock' });

    return options;
  }, [handleUnfollowClick, isFollowing]);

  return !isSelf ? (
    <>
      <Button
        scheme={!isFollowing ? ButtonSchemeEnum.PRIMARY : null}
        onClick={handleFollowClick}
        aria-label="follow-button"
        disabled={!isAuthenticated || isFollowing || isBlocking}
      >
        {isFollowing ? 'Following' : 'Follow'}
      </Button>
      <div className="last" style={{ marginLeft: 'auto' }}>
        {isAuthenticated && (
          <Dropdown options={dropdownOptions}>
            <Button
              scheme={ButtonSchemeEnum.SQUARE}
              disabled={!isAuthenticated}
            >
              ...
            </Button>
          </Dropdown>
        )}
      </div>
    </>
  ) : (
    <></>
  );
}

const mapStateToProps = (state: RootState) => ({
  user: selectCurrentProfileUser(state),
});

const mapDispatchToProps = {
  dispatchFollow: ProfileActions.follow,
  dispatchUnfollow: ProfileActions.unfollow,
  dispatchBlock: ProfileActions.block,
  dispatchAddSquad: ProfileActions.addSquad,
  dispatchRemoveSquad: SettingsActions.deleteSquad,
};

type ConnectedProps = ReturnType<typeof mapStateToProps> &
  typeof mapDispatchToProps;

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, memo)(FollowBlock) as React.ElementType;
