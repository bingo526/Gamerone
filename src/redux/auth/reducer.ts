import produce from 'immer';
import {
  LOGIN_SUCCESS,
  SIGNUP_SUCCESS,
  LOGOUT,
  CHECK_AUTHORIZATION_SUCCESS,
  UPDATE_CURRENT_USER,
  AuthActionTypes,
  AuthState,
} from './types';
import { getToken } from 'utils/token';
import { UserModel } from 'models/user';
import { INIT_STATE } from 'redux/types';

// TODO: Verify initial token
export const initState: AuthState = {
  idToken: getToken().get('idToken'),
  user: null,
};

export default function authReducer(
  state = initState,
  action: AuthActionTypes,
) {
  return produce(state, (draft) => {
    switch (action.type) {
      case INIT_STATE:
        draft.idToken = null;
        draft.user = null;
        break;
      case SIGNUP_SUCCESS:
      case LOGIN_SUCCESS: {
        const { token, user } = action.payload;
        draft.idToken = token || null;
        draft.user = new UserModel().fromDto(user);
        break;
      }

      case LOGOUT:
        draft.idToken = null;
        break;

      case CHECK_AUTHORIZATION_SUCCESS: {
        const { token, user } = action.payload;

        draft.idToken = token || null;
        draft.user = new UserModel().fromDto(user);
        break;
      }

      case UPDATE_CURRENT_USER:
        draft.user = new UserModel().fromDto(action.payload);
        break;

      default:
    }
  });
}
