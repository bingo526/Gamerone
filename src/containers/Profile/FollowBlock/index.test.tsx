import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import { AuthProvider } from 'provider/auth';
import { store } from 'redux/store';
import FollowBlock from './index';

const renderFollowBlock = (follows, dispatches = {}) => {
  const utils = render(
    <Provider store={store}>
      <AuthProvider isAuthenticated>
        <FollowBlock follows={follows} {...dispatches}></FollowBlock>
      </AuthProvider>
    </Provider>,
  );
  return utils;
};

describe('render FollowBlock', () => {
  test('should render the block and match the snapshot', () => {
    const { asFragment } = renderFollowBlock({
      followCount: 0,
      followerCount: 0,
      isFollowing: true,
      isBlocking: false,
    });

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('FollowBlock actions', () => {
  test('increases follower count on Follow click', () => {
    // TODO: Complete tests after block/unblock are done using profile reducer
  });
});
