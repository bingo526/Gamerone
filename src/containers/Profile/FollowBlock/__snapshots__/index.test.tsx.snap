// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render FollowBlock should render the block and match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center flex-row mt-3 mb-3"
  >
    <div
      class="d-flex flex-column mr-5"
    >
      <span
        aria-label="follower-count"
        style="font-size: 1.75rem; line-height: 0.75; font-weight: bold;"
      >
        0
      </span>
      <span
        style="font-size: 0.85rem;"
      >
        Followers
      </span>
    </div>
    <div
      class="d-flex flex-column"
    >
      <span
        aria-label="follow-count"
        style="font-size: 1.75rem; line-height: 0.75; font-weight: bold;"
      >
        0
      </span>
      <span
        style="font-size: 0.85rem;"
      >
        Following
      </span>
    </div>
  </div>
  <div
    class="d-flex flex-row mb-3"
  >
    <div
      class="flex-fill mr-2"
    >
      <button
        aria-label="follow-button"
        class="btn btn-primary"
        style="width: 100%;"
      >
        Follow
      </button>
    </div>
    <div>
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-primary dropdown-toggle"
          data-toggle="dropdown"
          id="dropdownMenuButton"
        >
          ...
        </button>
        <div>
          <div
            aria-labelledby="dropdownMenuButton"
            class="dropdown-menu"
          >
            <a
              class="dropdown-item"
              href="#block"
            >
              Block
            </a>
            <a
              class="dropdown-item"
              href="#report"
            >
              Report
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
